lib/poolex.ex

```diff
@@ -195,20 +189,29 @@ defmodule Poolex do
   """
   @spec run!(pool_id(), (worker :: pid() -> any()), list(run_option())) :: any()
   def run!(pool_id, fun, options \\ []) do
-    timeout = Keyword.get(options, :timeout, @default_wait_timeout)
-
-    {:ok, pid} = GenServer.call(pool_id, :get_idle_worker, timeout)
-
-    monitor_process = monitor_caller(pool_id, self(), pid)
+    checkout_timeout = Keyword.get(options, :checkout_timeout, @default_checkout_timeout)
+    {:ok, worker_pid} = get_idle_worker(pool_id, checkout_timeout)
+    monitor_process = monitor_caller(pool_id, self(), worker_pid)
 
     try do
-      fun.(pid)
+      fun.(worker_pid)
     after
       Process.exit(monitor_process, :kill)
-      GenServer.cast(pool_id, {:release_busy_worker, pid})
+      GenServer.cast(pool_id, {:release_busy_worker, worker_pid})
     end
   end
 
+  @spec get_idle_worker(pool_id(), timeout()) :: {:ok, worker()}
+  defp get_idle_worker(pool_id, checkout_timeout) do
+    caller_reference = make_ref()
+    GenServer.call(pool_id, {:get_idle_worker, caller_reference}, checkout_timeout)
+  catch
+    :exit,
+    {:timeout, {GenServer, :call, [_pool_id, {:get_idle_worker, caller_reference}, _timeout]}} ->
```

Comment:
I did this processing for two reasons:
1. I want to cover the logic with the case with a timeout from waiting for the GenServer.
2. In catch, I need to get `caller_refernce` to tell the pool to cancel the wait from the caller.

Reply:
You don't need the catch to get the `caller_reference`. In `:poolboy` is done slightly different: https://github.com/devinus/poolboy/blob/master/src/poolboy.erl#L61

You can get inspiration from it.


```elixir
  defp get_idle_worker(pool_id, checkout_timeout) do
    caller_reference = make_ref()
    try do
      GenServer.call(pool_id, {:get_idle_worker, caller_reference}, checkout_timeout)
    catch
      _, exception ->
         GenServer.cast(pool_id, {:cancel_waiting, caller_reference})
         reraise exception, __STACKTRACE__
    end
  end
```

Reply:
I want to keep the `{:error, :checkout_timeout}` answer from the `run` functions. This seems to me quite visual, even if we make one common `run` interface instead of two existing ones. I find it odd to handle `:exit, {:timeout...` on the call site.

What do you think about this implementation?

```elixir
  @spec get_idle_worker(pool_id(), timeout()) :: worker()
  defp get_idle_worker(pool_id, checkout_timeout) do
    caller_reference = make_ref()

    try do
      GenServer.call(pool_id, {:get_idle_worker, caller_reference}, checkout_timeout)
    catch
      :exit,
      {:timeout, {GenServer, :call, [_pool_id, {:get_idle_worker, ^caller_reference}, _timeout]}} ->
        raise CheckoutTimeoutError

      _, exception ->
        reraise exception, __STACKTRACE__
    after
      GenServer.cast(pool_id, {:cancel_waiting, caller_reference})
    end
  end
```